// services/RestService.js

import axios from 'axios';
import useLocalStorage from '../hooks/useLocalStorage'
import {ErrorTablePageContext} from './../contexts/errorTablePage'
import React, {useContext, useState} from 'react'
import Send from './Send'

    const upload = async (table, value) => {
        
            const res = await axios.post(`${BASE_HOST}/api/${table}`, value, headers );
            if (res.status !== 200) {
                if (res.status === 400 || res.status === 403) {
                    const criticalError = await res.text();
                    if (criticalError)
                        return { criticalError };
                }

                return { criticalError: `Could not upload file because of error status: ${res.status}.` };
            }
            return await res.data.json();
        }




    const insert = async (name, values) => {
        return await Send('POST', `/api${name}`, values)
        
    }
    const all = async (name, loadParams) => {
        return await Send('GET', `/api${name}`, loadParams);
    }
    const remove =  async (name, id) => {
        return await Send('DELETE', `/api${name}/${id}`);
    }

    const update = async (name, id, values) => {
        return await Send('PUT', `/api${name}/${id}`, values);
    }

   const byKey = async (name, id) => {
        return await Send('GET', `/api${name}/${id}`);
    }
   const deepByKey = async (name) => {
        return await Send('GET', `/api${name}`);
    }


export {insert, all, remove, update, byKey, deepByKey}

//get

//update


//delete
