// user-panel/user-panel.js
import React, { useMemo, useContext, useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
// import { useHistory } from "react-router-dom";
import ContextMenu, { Position } from 'devextreme-react/context-menu';
import List from 'devextreme-react/list';
// import { useAuth } from '../../contexts/auth';
import './user-panel.scss';
import useLocalStorage from '../hooks/useLocalStorage'
import {CurrentUserContext} from '../contexts/currentUser'

export default function UserPanel({ menuMode }) {
  // const { user, signOut } = useAuth();
  // const history = useHistory();
  const [token, setToken] = useLocalStorage('token');
  const [userInfo,setUserInfo] = useLocalStorage('loginUser')
  const navigate = useNavigate()
  const [state, dispatch] = useContext(CurrentUserContext);
  const [name, setName] = useState(null)


  const signOut = ()=>{
    setToken('null')
    dispatch({type: 'SET_UNAUTHORIZED'})
    navigate('/login')
  }
  function navigateToProfile() { 
    navigate("/profile");
  }
  const menuItems = useMemo(() => ([
    {
      text: 'Profile',
      icon: 'user',
      onClick: navigateToProfile
    },
    {
      text: 'Logout',
      icon: 'runner',
      onClick: signOut 
    }
  ]), []);

  return (
    <div className={'user-panel'}>
      <div className={'user-info'}>
        <div className={'image-container'}>
          <div
            style={{
              background:'url(C:\Users\Admin\Documents\Видео-статьи\Учёба\PlugAndPlay\new-app\src\node_modules\components\Forms\images\member1.png) no-repeat #fff',
              backgroundSize: 'cover'
            }}
            className={'user-image'} >
                <img src="\img\member1.png" className="img-responsive" alt="" />
              
              </div>
        </div>
        <div className={'user-name'}>{JSON.parse(userInfo).firstName}</div>
      </div>

      {menuMode === 'context' && (
        <ContextMenu
          items={menuItems}
          target={'.user-button'}
          showEvent={'dxclick'}
          width={210}
          cssClass={'user-menu'}
        >
          <Position my={'top center'} at={'bottom center'} />
        </ContextMenu>
      )}
      {menuMode === 'list' && (
        <List className={'dx-toolbar-menu-action'} items={menuItems} />
      )}
    </div>
  );
}
