// pages/routes.js
import {Switch, Route, Routes, Navigate} from "react-router-dom"
import GlobalFeed from "pages/globalFeed"
import Article from "pages/article"
import Authentication from "./authentication"
//import Authentication from "./authentification2/authentificationSecond"
// import TablePage from "../components/tablePage/tablePage"
import React, { useState, createContext, useContext, useEffect } from 'react';
import * as FaIcons from "react-icons/fa"
import * as AiIcons from "react-icons/ai"
import * as MdIcons from "react-icons/md"
import * as RiIcons from "react-icons/ri"
import * as GrIcons from "react-icons/gr"
import * as BsIcons from "react-icons/bs"
import DxTablePage from '../components/tablePage/dxTablePage/dxTablePage'
import PersonForm from '../components/Forms/PersonForms'
import TradeUnionForm from '../components/Forms/TradeUnionForm'
import UserForm from '../components/Forms/UserForm'
import {RequireAuth} from '../hoc/RequireAuth'
import {RequireUnAuth} from '../hoc/RequireUnAuth'
import Faq from '../components/faq_/faqs'
import About from '../pages/aboutUs/aboutUs'
import Payments from '../pages/payments/payments'

    export const navigation = [
        {
          text: 'Queries', icon: <MdIcons.MdQueryStats/>, cName: 'nav-text', path: '/admin/query', iconClosed:<RiIcons.RiArrowDownSFill/>,  iconOpened:<RiIcons.RiArrowUpSFill/>,
          items: [
            { id:1, text: "Lead's Flows", path: '/admin/query/lead_flow', icon: <GrIcons.GrContact/>, cName: 'nav-text' },
            { id:2,  text: "To Contact", path: '/admin/query/to_contact', icon: <AiIcons.AiOutlineMessage/>, cName: 'nav-text' }]
        },
        {
          text: 'Tables', icon: <AiIcons.AiOutlineTable/>, cName: 'nav-text',path: '/admin/tables', iconClosed:<RiIcons.RiArrowDownSFill/>,  iconOpened:<RiIcons.RiArrowUpSFill/>,
          items: [
            { id:3, text: 'tradeunion', path: '/admin/tables/tradeunion', icon: <AiIcons.AiOutlineTrademark/>, cName: 'nav-text' },
            { id:4, text: 'persons', path: '/admin/tables/persons', icon: <BsIcons.BsFileEarmarkPerson/>, cName: 'nav-text'},
            { id:5, text: 'education', path: '/admin/tables/educations', icon: <MdIcons.MdCastForEducation/>, cName: 'nav-text'}

          ]
        }
      ];

      
    export const navigationDx = [
      {
        text: 'Payments', icon: 'money', cName: 'nav-text', path: '/admin/payments', iconClosed:<RiIcons.RiArrowDownSFill/>,  iconOpened:<RiIcons.RiArrowUpSFill/>, resource: 'queries',
        items: [
          {text: "Pay operations", path: '/admin/payments/operations ', icon: 'hierarchy', cName: 'nav-text' , resource: 'leadToFlows' }]
      },
      {
        text: 'Tables', icon: 'tableproperties', cName: 'nav-text',path: '/admin/tables', iconClosed:<RiIcons.RiArrowDownSFill/>,  iconOpened:<RiIcons.RiArrowUpSFill/>,
        items: [
          {  text: 'tradeunion', path: '/admin/tables/tradeunion', icon: 'mergecells', cName: 'nav-text' , resource: 'tradeunion'},
          {  text: 'persons', path: '/admin/tables/persons', icon:'group', cName: 'nav-text' , resource: 'persons'},
          // {  text: 'education', path: '/admin/tables/educations', icon: 'toolbox', cName: 'nav-text' , resource: 'education'},
          // {  text: 'docpayments', path: '/admin/tables/docpayments', icon: 'background', cName: 'nav-text',  resource: 'docpayments'}

        ]
      },
      // {
      //   text: 'Forms', icon: 'globe', cName: 'nav-text',path: '/admin/forms', iconClosed:<RiIcons.RiArrowDownSFill/>,  iconOpened:<RiIcons.RiArrowUpSFill/>,
      //   items: [
      //     {  text: 'tradeunion', path: '/admin/forms/tradeunion', icon: 'mergecells', cName: 'nav-text',  resource: 'tradeunionform'},
      //     {  text: 'persons', path: '/admin/forms/persons', icon:'group', cName: 'nav-text',  resource: 'personform'},
      //   ] 
      // },

      {
        text: 'Helps', icon: 'help', cName: 'nav-text',path: '/admin/helps', iconClosed:<RiIcons.RiArrowDownSFill/>,  iconOpened:<RiIcons.RiArrowUpSFill/>,
        items: [
          {  text: "About Us", path: '/admin/helps/about_us', icon: 'like' , cName: 'nav-text' , resource: 'toContact'},
          {  text: 'FAQ', path: '/admin/helps/faq', icon:'info', cName: 'nav-text',  resource: 'personform'},
        ] 
      }
      // {
      //   text: 'Forms', icon: <AiIcons.AiOutlineTable/>, cName: 'nav-text',path: '/admin/tables', iconClosed:<RiIcons.RiArrowDownSFill/>,  iconOpened:<RiIcons.RiArrowUpSFill/>,
      //   items: [
      //     {  text: 'tradeunion', path: '/admin/tables/tradeunion', icon: <AiIcons.AiOutlineTrademark/>, cName: 'nav-text' },
      //     {  text: 'persons', path: '/admin/tables/persons', icon: <BsIcons.BsFileEarmarkPerson/>, cName: 'nav-text'},
      //     {  text: 'education', path: '/admin/tables/educations', icon: <MdIcons.MdCastForEducation/>, cName: 'nav-text'}

      //   ]
      // }
    ];
      
     export const routes = [

        // {
          {path: '/admin/tables/tradeunion', component: DxTablePage, table: 'tradeunion', dataSource: '/union', columns: [
            
            {dataField: "id", caption: "Id", dataType: "string", valueExpr: "id", tableProps:{
              alignment: "Center",            
              editType: "dropdownedit"
            }, roleWhoNotSee:['USER', 'ADMIN']},
            {dataField: "nameUnion", caption: "Name", dataType: "string", sortOrder: "asc" , tableProps:{
              alignment: "Center",   
              editType: "dropdownedit"
            },roleWhoNotSee:[]},
            {dataField: "city", caption: "City", dataType: "string", tableProps:{
              alignment: "Left",
              editType: "dropdownedit"
            }, roleWhoNotSee:[]},
            {dataField: "address", caption: "Adress", dataType: "string", tableProps:{
              alignment: "Center"
            }, roleWhoNotSee:[]},
            {dataField: "updated", caption: "Update", dataType: "datetime", tableProps:{
              alignment: "Center"
            },roleWhoNotSee:['USER']}
          ]},

          {path: '/admin/tables/persons', component: DxTablePage, table: 'persons', dataSource: '/persons', columns: [

            {dataField: "id", caption: "Id", dataType: "string", valueExpr: "id", tableProps:{
              alignment: "Center"
            }, roleWhoNotSee: ['USER', 'ADMIN']},
          
            {dataField: "firstName", caption: "Name", dataType: "string", sortOrder: "asc", tableProps:{
              alignment: "Rigth"             
            },roleWhoNotSee:[] },
            {dataField: "lastName", caption: "LastName", dataType: "string" , tableProps:{
              alignment: "Left"            
            }, roleWhoNotSee:[]},
            {dataField: "patronymic", caption: "Patron", dataType: "string", tableProps:{
              alignment: "Left"        
            },roleWhoNotSee: ['USER']
          },
            {dataField: "birth", caption: "Birth", dataType: "date" , tableProps:{
              alignment: "Center"
            }, roleWhoNotSee: ['USER']},
            
            {dataField: "education", caption: "Education", dataType: "string" , tableProps:{
              alignment: "Center"
            },roleWhoNotSee: ['USER']},
            
             ]},

          {path: '/admin/tables/educations', component: DxTablePage, table: 'educations', dataSource: '/education', columns: [
            
            {dataField: "name", caption: "Name", dataType: "string", tableProps:{
              alignment: "Center"
            },roleWhoNotSee:[] }
            ]},

          {path: '/admin/tables/docpayments', component: DxTablePage, table: 'docpayments', dataSource: '/docpayments', columns: [
          
            {dataField: "startPay", caption: "Start Pay", dataType: "date" , tableProps:{
              alignment: "Center"
            }, roleWhoNotSee:[]},
            {dataField: "endPay", caption: "End Pay", dataType: "date" , tableProps:{
              alignment: "Center"
            },roleWhoNotSee:[]},
            {dataField: "nameUnion", caption: "Name Union", dataType: "string" , tableProps:{
              alignment: "Center"
            },roleWhoNotSee:[]}
            // {dataField: "pay", caption: "PAY",  tableProps:{
            //   alignment: "Center"
            // },roleWhoNotSee:[]}
          ]},

          {path: '/admin/payments/docpayments', component: DxTablePage, table: 'operations', dataSource: '/docpayments', columns: [
          
            {dataField: "startPay", caption: "Start Paysdd", dataType: "date" , tableProps:{
              alignment: "Center"
            }, roleWhoNotSee:[]},
            {dataField: "endPay", caption: "End Pay", dataType: "date" , tableProps:{
              alignment: "Center"
            },roleWhoNotSee:[]},
            {dataField: "nameUnion", caption: "Name Union", dataType: "string" , tableProps:{
              alignment: "Center"
            },roleWhoNotSee:[]},
            {dataField: "pay", caption: "PAY",  tableProps:{
              alignment: "Center",
              type: "buttons",
              width:110,
              class:"dx-command-edit dx-command-edit-with-icons",
              icon:"airplane"
            },roleWhoNotSee:[]}
          ]}


      ]

export default routes.map(route => {
  return { ...route, component: withNavigationWatcher(route.component, { table: route.table, columns: route.columns }) };
});

const NavigationContext = createContext({});
const useNavigation = () => useContext(NavigationContext);


function NavigationProvider(props) {
  const [navigationData, setNavigationData] = useState({});

  return (
    <NavigationContext.Provider value={{ navigationData, setNavigationData }}
      {...props}
    />
  );
}


function withNavigationWatcher(Component, compProps) {
  return function (props) {
    const { path } = props.match;
    const { setNavigationData } = useNavigation();

    useEffect(() => {
      setNavigationData({ currentPath: path });
    }, [path, setNavigationData]);

    return React.createElement(Component, { ...compProps, ...props });
  }
}


export const Routess = () => {


  return(
    <>
    {<Routes>
      

      <Route path = "/login" element={
         <RequireUnAuth>
          <Authentication/>
        </RequireUnAuth>
      }/>
      <Route path = "/register" element={
         <RequireUnAuth>
          <Authentication/>
        </RequireUnAuth>
      }/>  
        <Route path = "/"  element= {
          <RequireAuth>
            <GlobalFeed/>
          </RequireAuth>
        }/>
        <Route path = "/admin/payments/:tablename"  element= {
          <RequireAuth>
            <Payments/>
          </RequireAuth>
        }/>
          <Route path = "admin/helps/faq"  element= {
          <RequireAuth>
            <Faq/>
          </RequireAuth>
        }/>
      <Route path = "admin/helps/about_us"  element= {
          <RequireAuth>
            <About/>
          </RequireAuth>
        }/>
        <Route path = "/global"  element= {
          <RequireAuth>
            <GlobalFeed/>
          </RequireAuth>
        }/>
        <Route path = "/admin/tables" element={
          <RequireAuth>
            <DxTablePage/>
          </RequireAuth>}/>
        <Route path = "/admin/tables/:tablename" element={
          <RequireAuth>
            <DxTablePage/>
          </RequireAuth>}/>   
        <Route path = "/persons/all" element = {
          <RequireAuth>
          <DxTablePage/>
          </RequireAuth>}/>
        <Route path = "/admin/forms/persons/:id"  element= {
          <RequireAuth>
          <PersonForm/>
          </RequireAuth>}/>
        <Route path = "/admin/forms/union/:id"  element= {
          <RequireAuth>
          <TradeUnionForm/>
          </RequireAuth>}/>
        <Route path = "/profile"  element= {
          <RequireAuth>
          <UserForm/>
          </RequireAuth>}/>
        {/* <Navigate to={'/global'} /> */}
    </Routes>
    
}
    </>
)
}
export { NavigationProvider, useNavigation, withNavigationWatcher }
